lib/balanced_vrp_clustering.rb

```diff
@@ -153,29 +287,86 @@ def recompute_centroids
 
           # correct the distance_from_and_to_depot info of the new cluster with the average of the points
           if centroid[4][:duration_from_and_to_depot]
-            centroid[4][:duration_from_and_to_depot][index] = @clusters[index].data_items.map{ |d| d[4][:duration_from_and_to_depot][index] }.reduce(&:+) / @clusters[index].data_items.size.to_f
+            centroid[4][:duration_from_and_to_depot] = @clusters[index].data_items.map{ |d| d[4][:duration_from_and_to_depot] }.reduce(&:+) / @clusters[index].data_items.size.to_f
           end
         }
 
-        @iterations += 1
+        swap_a_centroid_with_limit_violation
+
+        @iteration += 1
+      end
+
+      def swap_a_centroid_with_limit_violation
+        @clusters_with_limit_violation.map.with_index.sort_by{ |arr, _i| -arr.size }.each{ |preferred_clusters, violated_cluster|
```

Comment:
question: if you don't need `index`, only `@clusters_with_limit_violation.map.sort_by{ |arr| -arr.size }` is maybe enough?

Reply:
index is used after sort_by by `.each{ |preferred_clusters, violated_cluster|`, it is ignored in the first sort_by to avoid doing array manipulation (e.g., `@clusters_with_limit_violation.map.with_index.sort_by{ |obj| obj[0].size }`)

afaik, this is the standard way to sort the original indices with respect to another metric.